### PR TITLE
issue: 933949  Added KVM virtualization detection.

### PR DESCRIPTION
**Pull request by serglelv**
KVM detection added.
@OphirMunk @igor-ivanov 
Signed-off-by: sergeyly <sergeyly@mellanox.com>

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1087/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/830/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1088/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/831/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1089/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/832/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1095/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1108/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1110/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1115/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1126/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1132/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1137/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1147/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1159/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1162/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1172/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1185/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1189/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1213/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1239/ for details (Mellanox internal link).

**Comment by serglelv** [comment]
Jenkins tests fixed. Please review.
@OphirMunk @igor-ivanov

File: src/vma/iomux/epoll_wait_call.cpp [hunk 1]
```diff
@@ -195,6 +195,10 @@ bool epoll_wait_call::_wait(int timeout)
 		ready_fds = orig_os_api.epoll_wait(m_epfd, m_p_ready_events, m_maxevents, timeout);
 	}
 
+	if (is_suspend_required()) {
+		usleep(1);
```
Comment (OphirMunk) [inline src/vma/iomux/epoll_wait_call.cpp]: @serglelv why do you need to sleep for 1us? Why introducing a system call here?

File: src/vma/util/sys_vars.cpp [hunk 2]
```diff
@@ -336,6 +336,41 @@ void mce_sys_var::read_env_variable_with_pid(char* mce_sys_name, size_t mce_sys_
 	}
 }
 
+bool mce_sys_var::check_cpuinfo_flag(const char* flag)
```
Comment (OphirMunk) [inline src/vma/util/sys_vars.cpp]: @serglelv please explain why the flag is_hypervisor is a sufficient criteria for not calling the assembly code

File: src/vma/iomux/epoll_wait_call.cpp (hunk 1 above)
Comment (serglelv) [inline src/vma/iomux/epoll_wait_call.cpp]: For multi threads scenario the usleep(1) is used to suspend the current thread and provide right ordering in the epfd_info::increase_ring_ref_count_no_lock() method. The issue was reproduced only for KVM.

File: src/vma/util/sys_vars.cpp (hunk 2 above)
Comment (serglelv) [inline src/vma/util/sys_vars.cpp]: The is_hypervisor flag is used to fix the illegal instruction error in the instruction movntdq xmm0, register to the BlueFlame page pointed by the AX register. Please note the COPY_64B works fine in KVM and the problem is only accessing the BF page. The is_hypervisor flag is used for alternative implementation of the COPY_64B in VM. The flag also can be used to fix specific bugs for VM as well.

Comment (OphirMunk) [inline src/vma/util/sys_vars.cpp]: @serglelv If you tx 512 Bytes (where BlueFlame is not involved) - is there illegal instruction or not?

File: src/vma/iomux/epoll_wait_call.cpp (hunk 1 above)
Comment (OphirMunk) [inline src/vma/iomux/epoll_wait_call.cpp]: Per sergey this solves another issue: #902190

**Comment by OphirMunk** [comment]
Closed due to new root cause suggested by @serglelv 

The root cause of the issue is the value of MSL (Maximum Segment Lifetime).
Currently MSL is 60 sec.
tcp_impl.h L. 130 
#define TCP_MSL 60000UL /* The maximum segment lifetime in milliseconds */
With the current MSL ~ 100 connections are active in TIME_WAIT state and eventually the new connection cannot be created and test failed.
The test tcpOpenCloseClient.py/tcpOpenCloseserver.py (without any modifications) was successfully passed for 
MSL 1000 milliseconds i.e. #define TCP_MSL 1000UL /* The maximum segment lifetime in milliseconds */
It is interesting but without VMA (Linux sockets) the test was passed and MSL was ~800 milliseconds.
TIME-WAIT 0 0 12.12.12.13:5005 12.12.12.12:57258 timer:(timewait,735ms,0) 
BTW, in Linux MSL is configurable.

It seems the MSL parameter should be configurable in VMA.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1379/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/916/ for details (Mellanox internal link).

File: src/vma/util/sys_vars.cpp [hunk 3]
```diff
@@ -336,6 +336,41 @@ void mce_sys_var::read_env_variable_with_pid(char* mce_sys_name, size_t mce_sys_
...
+	char *line, *pch = NULL;
+	FILE *fp = NULL;
+	int line_size = MAX_CMD_LINE;
+
+	fp = fopen("/proc/cpuinfo", "r");
+	if (!fp) {
+		vlog_printf(VLOG_ERROR, "error while fopen\n");
+		print_vma_load_failure_msg();
+		exit(1);
+	}
```
Comment (OphirMunk) [inline src/vma/util/sys_vars.cpp]: @serglelv please remove the following 2 lines (no reason to exit() if cpuinfo cannot be read)
```
print_vma_load_failure_msg();
exit(1);
```
Instead just return false
```
return false;
```

File: src/vma/util/sys_vars.cpp [hunk 4]
```diff
@@ -336,6 +336,41 @@ void mce_sys_var::read_env_variable_with_pid(char* mce_sys_name, size_t mce_sys_
...
+	line = (char *)malloc(line_size);
+	BULLSEYE_EXCLUDE_BLOCK_START
+	if (!line) {
+		vlog_printf(VLOG_ERROR, "error while malloc\n");
+		print_vma_load_failure_msg();
+		exit(1);
+	}
```
Comment (OphirMunk) [inline src/vma/util/sys_vars.cpp]: Same comment as above: do not exit(). Instead just return false

File: src/vma/dev/qp_mgr.cpp [hunk 5]
```diff
@@ -587,8 +592,13 @@ void qp_mgr::mlx5_send(vma_ibv_send_wr *p_send_wqe)
 	 * implementations may use move-string-buffer assembler instructions,
 	 * which do not guarantee order of copying.
 	 */
+	if (likely(!m_b_is_hypervisor)) {
```
Comment (OphirMunk) [inline src/vma/dev/qp_mgr.cpp]: Must we have an if statement on the critical path? Please note the following link: https://gcc.gnu.org/onlinedocs/gcc-4.9.0/gcc/Function-Multiversioning.html 
Can you use Function Multi Versioning for SSE3 targets?

Comment (serglelv) [inline src/vma/dev/qp_mgr.cpp]: The issue is not related to the sse3 itself sse3 works fine. The problem is accessing to the bf using sse in HV.

File: src/vma/iomux/epoll_wait_call.cpp (hunk 1 above)
Comment (serglelv) [inline src/vma/iomux/epoll_wait_call.cpp]: Agree. It should be removed from the current pr.

File: src/vma/util/sys_vars.cpp (hunk 3 above)
Comment (serglelv) [inline src/vma/util/sys_vars.cpp]: It can be changed

**Comment by ilansmith** [comment]
Merged #304.

@serglelv - Please note that only the virtualization detection capability has been merged. It's incorporation with post_send functionality will be modified and used in PR #418.
Any other functionality introduced in this PR should be re-done in a separate PR.

**Pull request closed without merging**